proposals/scopes.md

```diff
@@ -196,106 +196,179 @@ interface OriginalPosition {
-    * Else if M >= 0, then
-        * M is used as an index into `names` field
-        * Note: The variable is accessible by evaluating the value expression for the entirety of this generated range.
-    * Else,
-      * Note: there are at least 2 sub-ranges.
-      * DATA offset into `names` field or -1
-        * Note: The variable is accessible using this value expression starting from the beginning of the generated range until the start of the next sub-range.
-        * Note: Use -1 to indicate that the current variable is unavailable (e.g. due to shadowing) in this sub-range.
-      * (M - 1) times
-        * DATA line in the generated code
-          * Note: The line is relative to the previous sub-range line or the start of the current generated range item if it’s the first for this loop.
-        * DATA column in the generated code
-          * Note: The column is relative to the column of the previous sub-range if it’s on the same line or absolute if it’s on a new line.
-        * DATA offset into `names` field or -1
-          * Note: The expression to obtain the current variable’s value in the sub-range that starts at line/column and ends at either the next sub-range start or this generated range’s end.
-          * Note: Use -1 to indicate that the current variable is unavailable (e.g. due to shadowing) in this sub-range.
-
-#### End Generated Range
-
-* DATA column in the generated code `**`
-  * Note: This is the point in generated code where the range ends. The line is the number of `;` preceding this item plus one.
-  * Note: The column is relative to the column of the previous item on the same line or absolute if there is no such item.
+
+`original_scope_variables` is a list of indices into the `names` array of the source map JSON. The list describes the original names
+of the scope's variables.
+
+To reduce the number of bytes required to encode the "scopes" information, we use relative values where possible:
+
+  * `sNAME` in `original_scope_start` is relative to the previous occurrence of `sNAME` (or absolute for the first).
+  * `sKIND` in `original_scope_start` is relative to the previous occurrence of `sKIND` (or absolute for the first).
+  * `uLINE` in `original_scope_start` and `original_scope_end` are relative to the previous occurrence (or absolute for the first).
+    This means a `uLINE` of a `original_scope_start` is relative to either its parents' start line or its preceding siblings' end line.
+  * `sVARIABLE` in `original_scope_variables` is relative to the previous occurrence of `sVARIABLE` (or absolute for the first).
+
+Each top-level `original_scope_tree` resets the "relative state". That is, each top-level `original_scope_tree` is decoded as if its the first.
+
+
+#### Generated Range Trees
+
+```
+generated_range_tree :=
+  generated_range_start
+  generated_range_callsite?
+  generated_range_bindings?
+  generated_range_subrange_binding?
+  generated_range_tree?
```

Comment:
There can be multiple `generated_range_tree` children.

Reply:
Same treatment as original_scope_tree (see above).